Add ability to specify gen3d forcefield and steps

psavery opened the pull request:
There are two ways to do this:

1. Using `oc.import_structure()`. If the structure does not have 3D
   coordinates, the forcefield and steps options may be set with the
   arguments `gen3d_forcefield` and `gen3d_steps`, respectively.
2. If a molecule does not have 3D coordinates, but is already in the
   database, 3D coordinates may be generated via
   `mol.structure.generate_3d()`, and `forcefield` and `steps` options
   can be passed to this function as well.

Note that `generate_3d()` can only be ran once per molecule in the database.
It cannot be ran for molecules that already have 3D coordinates.

psavery committed:
Add ability to specify gen3d forcefield and steps

There are two ways to do this:

1. Using oc.import_structure(). If the structure does not have 3D
   coordinates, the forcefield and steps options may be set with the
   arguments "gen3d_forcefield" and "gen3d_steps", respectively.
2. If a molecule does not have 3D coordinates, but is already in the
   database, 3D coordinates may be generated via
   mol.structure.generate_3d(), and "forcefield" and "steps" options
   can be passed to this function as well.

Note that generate_3d can only be ran once per molecule in the database.
It cannot be ran for molecules that already have 3D coordinates.

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

cjh1 commented on openchemistry/_data.py:
```diff
@@ -79,6 +79,24 @@ def __init__(self, cjson, molecule_id):
+                resp = GirderClient().get('molecules/%s/cjson' % self._id)
+            except HttpError as e:
+                fail_str = 'Molecule does not have 3D coordinates.'
+                if fail_str in e.responseText:
```
I'm not a huge fan of using the error text. May be the `molecules/%s/cjson` endpoint should return an empty object or `None` if there are no 3D coordinates?

psavery replied:
Okay. Yeah I think we could do that. The relevant code block is [here](https://github.com/OpenChemistry/mongochemserver/blob/8d5873ad8290e7cf345aff8cb43b49fec3342ee9/girder/molecules/molecules/molecule.py#L388-L392).

psavery replied:
Fixed in openchemistry/mongochemserver#173 and the latest commit here.

psavery committed:
Set output of cjson request directly on _cjson_

As of openchemistry/mongochemserver#173, None is now returned
rather than an exception being raised when there are no 3D
coordinates. If None gets returned, then it will be set on
_cjson_ and returned.

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

psavery committed:
Merge branch 'master' into gen3d-forcefield-steps

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

psavery commented:
I fixed the merge conflict.
